code/Modules/Selection/Dragger.cs

```diff
@@ -0,0 +1,139 @@
+        private WeldJoint holdJoint;
+        
+        public Dragger(MainPlayer player) : base(player)
+        {
+            Activate();
+        }
+
+        public IDraggable Dragged { get; private set; }
+        public PhysicsBody HeldBody => Dragged.Body;
+        public Rotation HeldRot { get; private set; }
+        public Entity HeldEntity { get; private set; }
+
+        public void Drag()
+        {
+            if (null == Dragged)
+            {
+                Entity entity = GetTraceResultEntity();
+                if (entity is not IDraggable draggable) return;
+                if (draggable.IsDraggable(player))
+                {
+                    StartDrag(entity, player.EyePos + player.EyeRot.Forward * HoldDistance, player.EyeRot);
+                }
+            }
+            else
+            {
+                GrabMove(player.EyePos, player.EyeRot.Forward, player.EyeRot);
+            }
+        }
+
+        public void StopDrag()
+        {
+            if (holdJoint.IsValid)
+            {
+                holdJoint.Remove();
+            }
+
+            if (Dragged != null)
+            {
+                HeldBody.EnableAutoSleeping = true;
+                Dragged.OnDragStop(player);
+            }
+
+            if (HeldEntity.IsValid())
+            {
+                Client client = player.GetClientOwner();
+                client?.Pvs.Remove( HeldEntity );
+            }
+
+            Dragged = null;
+            HeldRot = Rotation.Identity;
+            HeldEntity = null;
+        }
+
+        private void StartDrag(Entity entity, Vector3 grabPos, Rotation grabRot)
+        {
+            if (entity is not IDraggable draggable)
+                return;
+
+            if (!draggable.Body.IsValid())
+                return;
+
+            if (draggable.Body.PhysicsGroup == null)
+                return;
+
+            if (null != Dragged)
+                return;
+
+            StopDrag();
+
+            Dragged = draggable;
+            
+            HeldRot = grabRot.Inverse * HeldBody.Rotation;
+
+            holdBody.Position = grabPos;
+            holdBody.Rotation = HeldBody.Rotation;
+
+            HeldBody.Wake();
+            HeldBody.EnableAutoSleeping = false;
+            
+            holdJoint = PhysicsJoint.Weld
+                .From(holdBody)
+                .To(HeldBody, HeldBody.LocalMassCenter)
+                .WithLinearSpring(LinearFrequency, LinearDampingRatio, 0.0f)
+                .WithAngularSpring(AngularFrequency, AngularDampingRatio, 0.0f)
+                .Breakable(HeldBody.Mass * BreakLinearForce, 0)
+                .Create();
+
+            HeldEntity = entity;
+            Client client = player.Owner.GetClientOwner();
+            client?.Pvs.Add( HeldEntity );
+            
+            draggable.OnDragStart(player);
+        }
+
+        private void GrabMove(Vector3 startPos, Vector3 dir, Rotation rot)
+        {
+            if (!HeldBody.IsValid())
```

Comment:
question: whats diff heldBody and holdBody

Reply:
HeldBody is the body currently dragged, holdBody is the invisible physical object used to create the joint (the joint being the thing that allows us to drag items)